From da6344c265a2087a5171588d5c291bf3cd63daaa Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Sat, 22 Nov 2014 23:56:49 -0500
Subject: [md] dm: update wait_on_bit calls for RHEL

Message-id: <1416700613-25142-10-git-send-email-snitzer@redhat.com>
Patchwork-id: 100845
O-Subject: [RHEL7.1 PATCH 09/13] dm: update wait_on_bit calls for RHEL
Bugzilla: 1165246
RH-Acked-by: Joe Thornber <thornber@redhat.com>

BZ: 1165246

RHEL7 doesn't have upstream's wait_on_bit() interface change that
eliminates the action function parameter.

Implement wait_schedule_interruptible() so that dm_suspend() and
dm_resume()'s calls to wait_on_bit() are interruptible.

Signed-off-by: Mike Snitzer <snitzer@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/dm.c b/drivers/md/dm.c
index 6ac1e45..4a40395 100644
--- a/drivers/md/dm.c
+++ b/drivers/md/dm.c
@@ -2908,6 +2908,12 @@ static int __dm_suspend(struct mapped_device *md, struct dm_table *map,
  return r;
 }
 
+static int wait_schedule_interruptible(void *ptr)
+{
+ schedule();
+ return signal_pending(current);;
+}
+
 /*
  * We need to be able to change a mapping table under a mounted
  * filesystem.  For example we might want to move some data in
@@ -2940,7 +2946,8 @@ retry:
  if (dm_suspended_internally_md(md)) {
   /* already internally suspended, wait for internal resume */
   mutex_unlock(&md->suspend_lock);
-  r = wait_on_bit(&md->flags, DMF_SUSPENDED_INTERNALLY, TASK_INTERRUPTIBLE);
+  r = wait_on_bit(&md->flags, DMF_SUSPENDED_INTERNALLY,
+    wait_schedule_interruptible, TASK_INTERRUPTIBLE);
   if (r)
    return r;
   goto retry;
@@ -2998,7 +3005,8 @@ retry:
  if (dm_suspended_internally_md(md)) {
   /* already internally suspended, wait for internal resume */
   mutex_unlock(&md->suspend_lock);
-  r = wait_on_bit(&md->flags, DMF_SUSPENDED_INTERNALLY, TASK_INTERRUPTIBLE);
+  r = wait_on_bit(&md->flags, DMF_SUSPENDED_INTERNALLY,
+    wait_schedule_interruptible, TASK_INTERRUPTIBLE);
   if (r)
    return r;
   goto retry;
-- 
1.7.1